pkg/rc/replication_controller.go

```diff
@@ -0,0 +1,231 @@
+package rc
+
+import (
+	"strings"
+	"time"
+
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/kp/rcstore"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/pods"
+	"github.com/square/p2/pkg/util"
+)
+
+const (
+	rcIdLabel  = "replication_controller_id"
+	podIdLabel = "pod_id"
+)
+
+type ReplicationController interface {
+	ID() string
+	Manifest() *pods.Manifest
+	NodeSelector() labels.Selector
+	PodLabels() labels.Set
+	Disabled() bool
+	DesiredReplicas() int
+	CurrentNodes() ([]string, error)
+
+	// WatchDesires() spawns a goroutine that continually watches the desired replica count
+	// of this replication controller.
+	// Upon seeing any changes, it will converge observed state to desired state.
+	// The goroutine continues running until a value is received on the quitAndAck channel.
+	// The goroutine signals it has stopped by sending back a value on quitAndAck.
+	WatchDesires(quitAndAck chan struct{}) <-chan error
+}
+
+// I just copied this from kp.Store because it was a huge interface and I really only care about this one method!
+type kpStore interface {
+	SetPod(key string, manifest pods.Manifest) (time.Duration, error)
+}
+
+type replicationController struct {
+	rcstore.ReplicationControllerData
+
+	logger logging.Logger
+
+	kpStore       kpStore
+	rcStore       rcstore.ReplicationControllerStore
+	scheduler     Scheduler
+	podApplicator labels.Applicator
+}
+
+// replicationController implements the ReplicationController interface.
+var _ ReplicationController = &replicationController{}
+
+func New(
+	rcData rcstore.ReplicationControllerData,
+	kpStore kpStore,
+	rcStore rcstore.ReplicationControllerStore,
+	scheduler Scheduler,
+	podApplicator labels.Applicator,
+) *replicationController {
+	return &replicationController{
+		ReplicationControllerData: rcData,
+
+		logger:        logging.DefaultLogger,
+		kpStore:       kpStore,
+		rcStore:       rcStore,
+		scheduler:     scheduler,
+		podApplicator: podApplicator,
+	}
+}
+
+func (rc *replicationController) WatchDesires(quitAndAck chan struct{}) <-chan error {
+	/*
+		!!! TODO: This API now feels a bit strange to me.
+		The only reason a replicationController cares about its rcStore is so that it can call Watch on it.
+		HOWEVER, wouldn't it be smarter to have ReplicationControllerData provide a Watch() function???
+		That way, replicationControllers don't actually care how they're stored!
+		Then the Watch() function can be removed from ReplicationControllerStore.
+	*/
+	desiresChannel, errChannel := rc.rcStore.Watch(rc.ID(), quitAndAck)
```

Comment:
Throwing comment here so I can easily see what this used to look like. I wonder if it'll catch the entire comment.